framework/aster-frame/src/arch/x86/tdx_guest.rs

```diff
@@ -87,3 +150,279 @@ fn handle_io(trapframe: &mut dyn TdxTrapFrame, ve_info: &tdcall::TdgVeInfo) -> b
+        None => return Err(MmioError::DecodeFailed),
+    }
+
+    return Ok(instr.len());
+}
+
+fn decode_instr(rip: usize) -> Result<Instruction, MmioError> {
+    if !is_kernel_vaddr(rip) {
+        return Err(MmioError::InvalidAddress);
+    }
+    let code_data = {
+        const MAX_X86_INSTR_LEN: usize = 15;
+        let mut data = [0u8; MAX_X86_INSTR_LEN];
+        unsafe {
+            core::ptr::copy_nonoverlapping(rip as *const u8, data.as_mut_ptr(), data.len());
+        }
+        data
+    };
+    let mut decoder = Decoder::with_ip(64, &code_data, rip as u64, DecoderOptions::NONE);
+    let mut instr = Instruction::default();
+    decoder.decode_out(&mut instr);
+    if instr.is_invalid() {
+        return Err(MmioError::InvalidInstruction);
+    }
+    Ok(instr)
+}
+
+fn decode_mmio(instr: &Instruction) -> (Option<InstrMmioType>, u64) {
+    match instr.code() {
+        // 0x88
+        Code::Mov_rm8_r8 => (Some(InstrMmioType::Write), 1),
+        // 0x89
+        Code::Mov_rm16_r16 => (Some(InstrMmioType::Write), 2),
+        Code::Mov_rm32_r32 => (Some(InstrMmioType::Write), 4),
+        Code::Mov_rm64_r64 => (Some(InstrMmioType::Write), 8),
+        // 0xc6
+        Code::Mov_rm8_imm8 => (Some(InstrMmioType::WriteImm), 1),
+        // 0xc7
+        Code::Mov_rm16_imm16 => (Some(InstrMmioType::WriteImm), 2),
+        Code::Mov_rm32_imm32 => (Some(InstrMmioType::WriteImm), 4),
+        Code::Mov_rm64_imm32 => (Some(InstrMmioType::WriteImm), 8),
+        // 0x8a
+        Code::Mov_r8_rm8 => (Some(InstrMmioType::Read), 1),
+        // 0x8b
+        Code::Mov_r16_rm16 => (Some(InstrMmioType::Read), 2),
+        Code::Mov_r32_rm32 => (Some(InstrMmioType::Read), 4),
+        Code::Mov_r64_rm64 => (Some(InstrMmioType::Read), 8),
+        // 0xa4
+        Code::Movsb_m8_m8 => (Some(InstrMmioType::Movs), 1),
+        // 0xa5
+        Code::Movsw_m16_m16 => (Some(InstrMmioType::Movs), 2),
+        Code::Movsd_m32_m32 => (Some(InstrMmioType::Movs), 4),
+        Code::Movsq_m64_m64 => (Some(InstrMmioType::Movs), 8),
+        // 0x0f 0xb6
+        Code::Movzx_r16_rm8 | Code::Movzx_r32_rm8 | Code::Movzx_r64_rm8 => {
+            (Some(InstrMmioType::ReadZeroExtend), 1)
+        }
+        // 0x0f 0xb7
+        Code::Movzx_r16_rm16 | Code::Movzx_r32_rm16 | Code::Movzx_r64_rm16 => {
+            (Some(InstrMmioType::ReadZeroExtend), 2)
+        }
+        // 0x0f 0xbe
+        Code::Movsx_r16_rm8 | Code::Movsx_r32_rm8 | Code::Movsx_r64_rm8 => {
+            (Some(InstrMmioType::ReadSignExtend), 1)
+        }
+        // 0x0f 0xbf
+        Code::Movsx_r16_rm16 | Code::Movsx_r32_rm16 | Code::Movsx_r64_rm16 => {
+            (Some(InstrMmioType::ReadSignExtend), 2)
+        }
+        _ => (None, 0),
+    }
+}
+
+/// Sets the shared bit in the page table entry and notifies VMM about the change.
+/// The notification happens using MapGPA hypercall.
+///
+/// # Safety
+///
+/// User must ensure the given gpa is valid.
```

Comment:
This statement about the safety requirement is inaccurate. First, the description should be about the entire range, not only the given GPA. Second, the word "valid" is inappropriate. It is dangerous to make _valid_ memory pages _shared_.

And more descriptions should be given regarding why this method is extremely risky. One particular (potentially-surprising) side effect is that this method erases any valid data in the range. So the caller should not assume that after the `unprotect` operation, the data is still there.

In addition, after giving some more thoughts, I think the `phys_pages` part of the method name is misleading. The method unprotects a range of GPAs, which may include both physical memory pages and MMIO ranges. So I think `unprotect_gpa_range` would be most accurate.

BTW. ChatGPT is your friend.